boxtree/fmm.py

```diff
@@ -20,234 +28,109 @@
-
-    mpole_result, timing_future = wrangler.eval_multipoles(
-            traversal.target_boxes_sep_smaller_by_source_level,
-            traversal.from_sep_smaller_by_level,
-            mpole_exps)
-
-    recorder.add("eval_multipoles", timing_future)
-
-    potentials = potentials + mpole_result
-
-    # these potentials are called beta in [1]
-
-    if traversal.from_sep_close_smaller_starts is not None:
-        logger.debug("evaluate separated close smaller interactions directly "
-                "('list 3 close')")
-
-        direct_result, timing_future = wrangler.eval_direct(
-                traversal.target_boxes,
-                traversal.from_sep_close_smaller_starts,
-                traversal.from_sep_close_smaller_lists,
-                src_weight_vecs)
-
-        recorder.add("eval_direct", timing_future)
-
-        potentials = potentials + direct_result
-
-    # }}}
-
-    # {{{ "Stage 6:" form locals for separated bigger source boxes ("list 4")
-
-    local_result, timing_future = wrangler.form_locals(
-            traversal.level_start_target_or_target_parent_box_nrs,
-            traversal.target_or_target_parent_boxes,
-            traversal.from_sep_bigger_starts,
-            traversal.from_sep_bigger_lists,
-            src_weight_vecs)
-
-    recorder.add("form_locals", timing_future)
-
-    local_exps = local_exps + local_result
-
-    if traversal.from_sep_close_bigger_starts is not None:
-        direct_result, timing_future = wrangler.eval_direct(
-                traversal.target_boxes,
-                traversal.from_sep_close_bigger_starts,
-                traversal.from_sep_close_bigger_lists,
-                src_weight_vecs)
-
-        recorder.add("eval_direct", timing_future)
-
-        potentials = potentials + direct_result
-
-    # }}}
-
-    # {{{ "Stage 7:" propagate local_exps downward
 
-    local_exps, timing_future = wrangler.refine_locals(
-            traversal.level_start_target_or_target_parent_box_nrs,
-            traversal.target_or_target_parent_boxes,
-            local_exps)
 
-    recorder.add("refine_locals", timing_future)
+class ExpansionWranglerInterface(ABC):
+    """Abstract expansion handling interface for use with :func:`drive_fmm`.
 
-    # }}}
+    See this
+    `test code <https://github.com/inducer/boxtree/blob/master/test/test_fmm.py>`_
+    for a very simple sample implementation.
 
-    # {{{ "Stage 8:" evaluate locals
+    .. note::
 
-    local_result, timing_future = wrangler.eval_locals(
-            traversal.level_start_target_box_nrs,
-            traversal.target_boxes,
-            local_exps)
+        Wranglers should not hold a reference (and thereby be specific to) a
+        :class:`boxtree.Tree` instance. Their purpose is to host caches for
+        generated translation code that is reusable across trees.
+        It is OK for expansion wranglers to be specific to a given kernel
+        (or set of kernels).
```

Comment:
Does this comment still apply to the latest version? Looks like now `ExpansionWranglerInterface` holds both tree dependent and tree independent data, so should we move this comment to `TreeIndependentDataForWrangler`?

Reply:
Good catch! 7604d83